File: api/src/main/java/com/ruby/rubia_server/core/controller/AudioController.java
package com.ruby.rubia_server.core.controller;

import com.ruby.rubia_server.core.entity.AudioMessage;
import com.ruby.rubia_server.core.repository.AudioMessageRepository;
import com.ruby.rubia_server.core.service.AudioProcessingService;
import com.ruby.rubia_server.core.service.AudioStorageService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.core.io.InputStreamResource;
import org.springframework.web.bind.annotation.*;

import java.io.InputStream;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Slf4j
@RestController
@RequestMapping("/api/audio")
@RequiredArgsConstructor
public class AudioController {

    private final AudioProcessingService audioProcessingService;
    private final AudioMessageRepository audioMessageRepository;
    private final AudioStorageService audioStorageService;

    @PostMapping("/send")
    public ResponseEntity<?> sendAudio(@RequestBody Map<String, String> request) {
        try {
            String toNumber = request.get("toNumber");
            String audioUrl = request.get("audioUrl");

            if (toNumber == null || audioUrl == null) {
                return ResponseEntity.badRequest().body(Map.of(
                    "error", "Missing required parameters: toNumber and audioUrl"
                ));
            }

            String messageId = audioProcessingService.sendAudio(toNumber, audioUrl);

            return ResponseEntity.ok(Map.of(
                "status", "success",
                "messageId", messageId
            ));

        } catch (Exception e) {
            log.error("Error sending audio: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("error", e.getMessage()));
        }
    }

    @GetMapping("/messages")
    public ResponseEntity<List<AudioMessage>> listMessages(
            @RequestParam(required = false) String fromNumber,
            @RequestParam(required = false) AudioMessage.ProcessingStatus status,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {

        PageRequest pageRequest = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<AudioMessage> messagePage;

        if (fromNumber != null) {
            messagePage = audioMessageRepository.findAll(pageRequest);
            List<AudioMessage> filteredMessages = messagePage.getContent().stream()
                    .filter(msg -> fromNumber.equals(msg.getFromNumber()))
                    .toList();
            return ResponseEntity.ok(filteredMessages);
        } else if (status != null) {
            List<AudioMessage> messages = audioMessageRepository.findByStatus(status);
            return ResponseEntity.ok(messages);
        } else {
            messagePage = audioMessageRepository.findAll(pageRequest);
            return ResponseEntity.ok(messagePage.getContent());
        }
    }

    @GetMapping("/messages/{id}")
    public ResponseEntity<?> getMessage(@PathVariable UUID id) {
        return audioMessageRepository.findById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/messages/{id}/download")
    public ResponseEntity<?> downloadAudio(@PathVariable UUID id) {
        try {
            Optional<AudioMessage> audioMessageOpt = audioMessageRepository.findById(id);
            if (audioMessageOpt.isEmpty()) {
                return ResponseEntity.notFound().build();
            }

            AudioMessage audioMessage = audioMessageOpt.get();
            if (audioMessage.getFilePath() == null) {
                return ResponseEntity.badRequest().body(Map.of(
                    "error", "Audio file not available"
                ));
            }

            Optional<InputStream> audioStream = audioStorageService.retrieve(audioMessage.getFilePath());
            if (audioStream.isEmpty()) {
                return ResponseEntity.notFound().build();
            }

            String fileName = audioMessage.getMessageId() + ".ogg";
            String mimeType = audioMessage.getMimeType() != null ? 
                audioMessage.getMimeType() : "audio/ogg";

            return ResponseEntity.ok()
                    .contentType(MediaType.parseMediaType(mimeType))
                    .header("Content-Disposition", "attachment; filename=\"" + fileName + "\"")
                    .body(new InputStreamResource(audioStream.get()));

        } catch (Exception e) {
            log.error("Error downloading audio: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("error", "Failed to download audio file"));
        }
    }

    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getStats() {
        long totalMessages = audioMessageRepository.count();
        long receivedCount = audioMessageRepository.countByStatus(AudioMessage.ProcessingStatus.RECEIVED);
        long processingCount = audioMessageRepository.countByStatus(AudioMessage.ProcessingStatus.DOWNLOADING) +
                              audioMessageRepository.countByStatus(AudioMessage.ProcessingStatus.PROCESSING);
        long completedCount = audioMessageRepository.countByStatus(AudioMessage.ProcessingStatus.COMPLETED);
        long failedCount = audioMessageRepository.countByStatus(AudioMessage.ProcessingStatus.FAILED);

        return ResponseEntity.ok(Map.of(
            "total", totalMessages,
            "received", receivedCount,
            "processing", processingCount,
            "completed", completedCount,
            "failed", failedCount
        ));
    }

    @DeleteMapping("/messages/{id}")
    public ResponseEntity<?> deleteAudioMessage(@PathVariable UUID id) {
        try {
            Optional<AudioMessage> audioMessageOpt = audioMessageRepository.findById(id);
            if (audioMessageOpt.isEmpty()) {
                return ResponseEntity.notFound().build();
            }

            AudioMessage audioMessage = audioMessageOpt.get();
            
            // Delete file if exists
            if (audioMessage.getFilePath() != null) {
                audioStorageService.delete(audioMessage.getFilePath());
            }

            // Delete database record
            audioMessageRepository.delete(audioMessage);

            return ResponseEntity.ok(Map.of("status", "deleted"));

        } catch (Exception e) {
            log.error("Error deleting audio message: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("error", "Failed to delete audio message"));
        }
    }
}

File: api-dotnet/Rubia.Server/Controllers/AudioController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Enums;
using Rubia.Server.Repositories;
using Rubia.Server.Services;

namespace Rubia.Server.Controllers;

[ApiController]
[Route("api/audio")]
[Authorize]
public class AudioController : ControllerBase
{
    private readonly IAudioProcessingService _audioProcessingService;
    private readonly IAudioMessageRepository _audioMessageRepository;
    private readonly IAudioStorageService _audioStorageService;
    private readonly ILogger<AudioController> _logger;

    public AudioController(
        IAudioProcessingService audioProcessingService,
        IAudioMessageRepository audioMessageRepository,
        IAudioStorageService audioStorageService,
        ILogger<AudioController> logger)
    {
        _audioProcessingService = audioProcessingService;
        _audioMessageRepository = audioMessageRepository;
        _audioStorageService = audioStorageService;
        _logger = logger;
    }

    [HttpPost("send")]
    public async Task<IActionResult> SendAudio([FromBody] SendAudioRequestDto request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.ToNumber) || string.IsNullOrEmpty(request.AudioUrl))
            {
                return BadRequest(new { error = "Missing required parameters: toNumber and audioUrl" });
            }

            var messageId = await _audioProcessingService.SendAudioAsync(request.ToNumber, request.AudioUrl);

            return Ok(new { status = "success", messageId });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error sending audio: {Error}", ex.Message);
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpGet("messages")]
    public async Task<ActionResult<IEnumerable<AudioMessageDto>>> ListMessages(
        [FromQuery] string? fromNumber,
        [FromQuery] ProcessingStatus? status,
        [FromQuery] int page = 0,
        [FromQuery] int size = 20)
    {
        try
        {
            IEnumerable<AudioMessage> messages;

            if (!string.IsNullOrEmpty(fromNumber))
            {
                messages = await _audioMessageRepository.GetByFromNumberAsync(fromNumber);
            }
            else if (status.HasValue)
            {
                messages = await _audioMessageRepository.GetByStatusAsync(status.Value);
            }
            else
            {
                messages = await _audioMessageRepository.GetAllAsync(page, size);
            }

            var dtos = messages.Select(MapToDto);
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error listing audio messages");
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("messages/{id:guid}")]
    public async Task<ActionResult<AudioMessageDto>> GetMessage(Guid id)
    {
        try
        {
            var message = await _audioMessageRepository.GetByIdAsync(id);
            if (message == null)
            {
                return NotFound();
            }

            return Ok(MapToDto(message));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audio message {Id}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("messages/{id:guid}/download")]
    public async Task<IActionResult> DownloadAudio(Guid id)
    {
        try
        {
            var audioMessage = await _audioMessageRepository.GetByIdAsync(id);
            if (audioMessage == null)
            {
                return NotFound();
            }

            if (string.IsNullOrEmpty(audioMessage.FilePath))
            {
                return BadRequest(new { error = "Audio file not available" });
            }

            var (stream, contentType, fileName) = await _audioStorageService.RetrieveAsync(audioMessage.FilePath);
            if (stream == null)
            {
                return NotFound();
            }

            var downloadFileName = $"{audioMessage.MessageId}.ogg";
            var mimeType = audioMessage.MimeType ?? "audio/ogg";

            return File(stream, mimeType, downloadFileName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error downloading audio {Id}", id);
            return StatusCode(500, new { error = "Failed to download audio file" });
        }
    }

    [HttpGet("stats")]
    public async Task<ActionResult<AudioStatsDto>> GetStats()
    {
        try
        {
            var totalMessages = await _audioMessageRepository.CountAsync();
            var receivedCount = await _audioMessageRepository.CountByStatusAsync(ProcessingStatus.RECEIVED);
            var processingCount = await _audioMessageRepository.CountByStatusAsync(ProcessingStatus.DOWNLOADING) +
                                 await _audioMessageRepository.CountByStatusAsync(ProcessingStatus.PROCESSING);
            var completedCount = await _audioMessageRepository.CountByStatusAsync(ProcessingStatus.COMPLETED);
            var failedCount = await _audioMessageRepository.CountByStatusAsync(ProcessingStatus.FAILED);

            var stats = new AudioStatsDto
            {
                Total = totalMessages,
                Received = receivedCount,
                Processing = processingCount,
                Completed = completedCount,
                Failed = failedCount
            };

            return Ok(stats);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting audio stats");
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpDelete("messages/{id:guid}")]
    public async Task<IActionResult> DeleteAudioMessage(Guid id)
    {
        try
        {
            var audioMessage = await _audioMessageRepository.GetByIdAsync(id);
            if (audioMessage == null)
            {
                return NotFound();
            }

            // Delete file if exists
            if (!string.IsNullOrEmpty(audioMessage.FilePath))
            {
                await _audioStorageService.DeleteAsync(audioMessage.FilePath);
            }

            // Delete database record
            await _audioMessageRepository.DeleteAsync(id);

            return Ok(new { status = "deleted" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting audio message {Id}", id);
            return StatusCode(500, new { error = "Failed to delete audio message" });
        }
    }

    private static AudioMessageDto MapToDto(AudioMessage audioMessage)
    {
        return new AudioMessageDto
        {
            Id = audioMessage.Id,
            MessageId = audioMessage.MessageId,
            FromNumber = audioMessage.FromNumber,
            ToNumber = audioMessage.ToNumber,
            Direction = audioMessage.Direction,
            AudioUrl = audioMessage.AudioUrl,
            FilePath = audioMessage.FilePath,
            MimeType = audioMessage.MimeType,
            DurationSeconds = audioMessage.DurationSeconds,
            FileSizeBytes = audioMessage.FileSizeBytes,
            Status = audioMessage.Status,
            CreatedAt = audioMessage.CreatedAt,
            ProcessedAt = audioMessage.ProcessedAt,
            ErrorMessage = audioMessage.ErrorMessage,
            ConversationId = audioMessage.ConversationId
        };
    }
}